### PR TITLE
Add logging

### PR DESCRIPTION
**Pull request by ryardley**
Closes: #124 

- [x] add tracing crate 
- [x] replace `println` with `tracing::xxxx()` where apprpriate

## Release Notes

- **New Features**
  - Enhanced logging capabilities across multiple components using structured logging with the `tracing` crate.
  - New asynchronous handling for ciphertext decryption in the Keyshare actor.
  - Added new methods for improved error handling and event management in various modules.

- **Bug Fixes**
  - Improved error handling in event processing to provide clearer messages and prevent duplicate event submissions.

- **Documentation**
  - Updated comments and documentation to reflect changes in method signatures and functionality.

- **Tests**
  - Expanded test cases to ensure robust event handling and data integrity across the system.
  - Added new tests for the P2P actor to validate event forwarding behavior.

### CODE REVIEW

File: packages/ciphernode/test_helpers/src/utils.rs [hunk 1]
```diff
@@ -15,7 +22,9 @@ pub fn write_file_with_dirs(path: &str, content: &[u8]) -> std::io::Result<()> {
     // Open the file (creates it if it doesn't exist) and write the content
     let mut file = fs::File::create(&abs_path)?;
     file.write_all(content)?;
-
-    println!("File written successfully: {:?}", abs_path);
+    trace!(
+        path = abs_path.to_str().unwrap(),
+        "File written successfully!"
+    );
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/test_helpers/src/utils.rs:28]: This is a test helper

File: packages/ciphernode/evm/src/enclave_sol_reader.rs [hunk 2]
```diff
@@ -55,21 +56,24 @@ fn extractor(data: &LogData, topic: Option<&B256>, chain_id: u64) -> Option<Encl
...
                 return None;
             };
             Some(EnclaveEvent::from(E3RequestedWithChainId(event, chain_id)))
         }
         Some(&IEnclave::CiphertextOutputPublished::SIGNATURE_HASH) => {
             let Ok(event) = IEnclave::CiphertextOutputPublished::decode_log_data(data, true) else {
-                println!("Error parsing event CiphertextOutputPublished"); // TODO: provide more info
+                error!("Error parsing event CiphertextOutputPublished after topic matched!"); // TODO: provide more info
                 return None;
             };
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/src/enclave_sol_reader.rs:68]: This is fine as is

File: packages/ciphernode/logger/src/logger.rs [hunk 3]
```diff
@@ -29,30 +28,12 @@ impl Actor for SimpleLogger {
...
-                    self.name, data.e3_id, data.threshold_m, data.seed
-                );
-            }
-            EnclaveEvent::EnclaveError { data, .. } => {
-                println!("[{}]: EnclaveError('{}')", self.name, data.message);
-            }
-            _ => println!("[{}]: {}", self.name, msg),
-        }
+        match msg {
+            EnclaveEvent::EnclaveError { .. } => error!(event=%msg, "ERROR!"),
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/logger/src/logger.rs:32]: This is a good point perhaps but there isn't much that a local user shouldn't really see in an error log at this point in our dev cycle

File: packages/ciphernode/logger/src/logger.rs [hunk 4]
```diff
@@ -29,30 +28,12 @@
...
+            _ => match msg.get_e3_id() {
+                Some(e3_id) => info!(me=self.name, evt=%msg, e3_id=%e3_id, "Event Broadcasted"),
+                None => info!(me=self.name, evt=%msg, "Event Broadcasted"),
+            },
+        };
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/logger/src/logger.rs:37]: It is slightly less verbose the way it is

File: packages/ciphernode/enclave/src/main.rs [hunk 5]
```diff
@@ -29,11 +30,12 @@ pub struct Args {
 
 #[actix_rt::main]
 async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    tracing_subscriber::fmt::init();
     println!("\n\n\n\n\n{}", OWO);
     println!("\n\n\n\n");
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/enclave/src/main.rs:35]: NOPE!!

File: packages/ciphernode/p2p/src/libp2p_router.rs [hunk 6]
```diff
@@ -146,30 +148,18 @@ impl EnclaveRouter {
...
                         message,
                     })) => {
-                        // println!(
-                        //     "Got message with id: {id} from peer: {peer_id}",
-                        // );
-                        // println!("{:?}", message);
+                        trace!(
+                            "Got message with id: {id} from peer: {peer_id}",
+                        );
+                        trace!("{:?}", message);
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/p2p/src/libp2p_router.rs:154]: Not at this point

File: packages/ciphernode/sortition/src/sortition.rs [hunk 7]
```diff
@@ -1,9 +1,12 @@
...
 use crate::DistanceSortition;
 use actix::prelude::*;
 use alloy::primitives::Address;
-use enclave_core::{CiphernodeAdded, CiphernodeRemoved, EnclaveEvent, EventBus, Seed, Subscribe};
+use anyhow::{anyhow, Result};
+use enclave_core::{
+    BusError, CiphernodeAdded, CiphernodeRemoved, EnclaveErrorType, EnclaveEvent, EventBus, Seed,
+    Subscribe,
+};
+use std::collections::HashSet;
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/sortition/src/sortition.rs:9]: Thisebds the error to the bus whuch prints it

**Review by hmzakhalid — APPROVED** [review]

**Pull request merged**